--- /dev/null
+++ b/me/milesglitch/client/Hooks.java
@@ -1,0 +_,12 @@
+package me.milesglitch.client;
+
+import net.fabricmc.fabric.api.renderer.v1.Renderer;
+import net.fabricmc.fabric.impl.client.indigo.Indigo;
+import net.fabricmc.fabric.impl.client.indigo.renderer.IndigoRenderer;
+
+public class Hooks {
+    public static void startClient(){
+        Indigo.LOGGER.info("[Indigo] Registering Indigo renderer!");
+        Renderer.register(IndigoRenderer.INSTANCE);
+    }
+}
