--- /dev/null
+++ b/me/milesglitch/client/VanillaRenderer.java
@@ -1,0 +_,32 @@
+package me.milesglitch.client;
+
+import net.fabricmc.fabric.api.renderer.v1.Renderer;
+import net.fabricmc.fabric.api.renderer.v1.material.MaterialFinder;
+import net.fabricmc.fabric.api.renderer.v1.material.RenderMaterial;
+import net.fabricmc.fabric.api.renderer.v1.mesh.MutableMesh;
+import net.minecraft.resources.ResourceLocation;
+import org.jetbrains.annotations.Nullable;
+
+public class VanillaRenderer implements Renderer {
+    public static final VanillaRenderer VANILLA_RENDERER = new VanillaRenderer();
+
+    @Override
+    public MutableMesh mutableMesh() {
+        return null;
+    }
+
+    @Override
+    public MaterialFinder materialFinder() {
+        return null;
+    }
+
+    @Override
+    public @Nullable RenderMaterial materialById(ResourceLocation id) {
+        return null;
+    }
+
+    @Override
+    public boolean registerMaterial(ResourceLocation id, RenderMaterial material) {
+        return false;
+    }
+}
