--- a/net/minecraft/client/renderer/GameRenderer.java
+++ b/net/minecraft/client/renderer/GameRenderer.java
@@ -14,6 +_,8 @@
 import java.nio.file.Path;
 import java.util.Locale;
 import javax.annotation.Nullable;
+
+import net.caffeinemc.mods.sodium.client.gui.console.ConsoleHooks;
 import net.minecraft.CrashReport;
 import net.minecraft.CrashReportCategory;
 import net.minecraft.ReportedException;
@@ -65,7 +_,9 @@
 import org.joml.Matrix4fStack;
 import org.joml.Quaternionf;
 import org.joml.Vector3f;
+import org.lwjgl.glfw.GLFW;
 import org.slf4j.Logger;
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
 
 public class GameRenderer implements AutoCloseable {
     private static final ResourceLocation BLUR_POST_CHAIN_ID = ResourceLocation.withDefaultNamespace("blur");
@@ -107,6 +_,7 @@
     private ResourceLocation postEffectId;
     private boolean effectActive;
     private final Camera mainCamera = new Camera();
+    private static boolean HAS_RENDERED_OVERLAY_ONCE = false;
 
     public GameRenderer(Minecraft minecraft, ItemInHandRenderer itemInHandRenderer, ResourceManager resourceManager, RenderBuffers renderBuffers) {
         this.minecraft = minecraft;
@@ -471,6 +_,7 @@
 
                 this.minecraft.gui.render(guiGraphics, deltaTracker);
                 guiGraphics.flush();
+                this.handler$zci000$sodium$onRender(deltaTracker, renderLevel, (CallbackInfo)null, profilerFiller);
                 RenderSystem.clear(256);
                 profilerFiller.pop();
             }
@@ -535,6 +_,7 @@
             }
 
             guiGraphics.flush();
+            this.handler$zci000$sodium$onRender(deltaTracker, renderLevel, (CallbackInfo)null, profilerFiller);
             modelViewStack.popMatrix();
             this.resourcePool.endFrame();
         }
@@ -747,5 +_,16 @@
 
     public OverlayTexture overlayTexture() {
         return this.overlayTexture;
+    }
+
+    private void handler$zci000$sodium$onRender(DeltaTracker deltaTracker, boolean bl, CallbackInfo ci, ProfilerFiller profiler) {
+        if (Minecraft.getInstance().getOverlay() == null || HAS_RENDERED_OVERLAY_ONCE) {
+            profiler.push("sodium_console_overlay");
+            GuiGraphics drawContext = new GuiGraphics(this.minecraft, this.renderBuffers.bufferSource());
+            ConsoleHooks.render(drawContext, GLFW.glfwGetTime());
+            drawContext.flush();
+            profiler.pop();
+            HAS_RENDERED_OVERLAY_ONCE = true;
+        }
     }
 }
