--- a/net/minecraft/client/renderer/RenderBuffers.java
+++ b/net/minecraft/client/renderer/RenderBuffers.java
@@ -3,6 +_,8 @@
 import com.mojang.blaze3d.vertex.ByteBufferBuilder;
 import it.unimi.dsi.fastutil.objects.Object2ObjectLinkedOpenHashMap;
 import java.util.SequencedMap;
+
+import net.caffeinemc.mods.sodium.client.render.chunk.NonStoringBuilderPool;
 import net.minecraft.Util;
 import net.minecraft.client.resources.model.ModelBakery;
 
@@ -14,23 +_,23 @@
     private final OutlineBufferSource outlineBufferSource;
 
     public RenderBuffers(int bufferCount) {
-        this.sectionBufferPool = SectionBufferBuilderPool.allocate(bufferCount);
+        this.sectionBufferPool = new NonStoringBuilderPool();
         SequencedMap<RenderType, ByteBufferBuilder> map = Util.make(new Object2ObjectLinkedOpenHashMap<>(), map2 -> {
             map2.put(Sheets.solidBlockSheet(), this.fixedBufferPack.buffer(RenderType.solid()));
             map2.put(Sheets.cutoutBlockSheet(), this.fixedBufferPack.buffer(RenderType.cutout()));
             map2.put(Sheets.bannerSheet(), this.fixedBufferPack.buffer(RenderType.cutoutMipped()));
             map2.put(Sheets.translucentItemSheet(), this.fixedBufferPack.buffer(RenderType.translucent()));
-            put((Object2ObjectLinkedOpenHashMap<RenderType, ByteBufferBuilder>)map2, Sheets.shieldSheet());
-            put((Object2ObjectLinkedOpenHashMap<RenderType, ByteBufferBuilder>)map2, Sheets.bedSheet());
-            put((Object2ObjectLinkedOpenHashMap<RenderType, ByteBufferBuilder>)map2, Sheets.shulkerBoxSheet());
-            put((Object2ObjectLinkedOpenHashMap<RenderType, ByteBufferBuilder>)map2, Sheets.signSheet());
-            put((Object2ObjectLinkedOpenHashMap<RenderType, ByteBufferBuilder>)map2, Sheets.hangingSignSheet());
+            put(map2, Sheets.shieldSheet());
+            put(map2, Sheets.bedSheet());
+            put(map2, Sheets.shulkerBoxSheet());
+            put(map2, Sheets.signSheet());
+            put(map2, Sheets.hangingSignSheet());
             map2.put(Sheets.chestSheet(), new ByteBufferBuilder(786432));
-            put((Object2ObjectLinkedOpenHashMap<RenderType, ByteBufferBuilder>)map2, RenderType.armorEntityGlint());
-            put((Object2ObjectLinkedOpenHashMap<RenderType, ByteBufferBuilder>)map2, RenderType.glint());
-            put((Object2ObjectLinkedOpenHashMap<RenderType, ByteBufferBuilder>)map2, RenderType.glintTranslucent());
-            put((Object2ObjectLinkedOpenHashMap<RenderType, ByteBufferBuilder>)map2, RenderType.entityGlint());
-            put((Object2ObjectLinkedOpenHashMap<RenderType, ByteBufferBuilder>)map2, RenderType.waterMask());
+            put(map2, RenderType.armorEntityGlint());
+            put(map2, RenderType.glint());
+            put(map2, RenderType.glintTranslucent());
+            put(map2, RenderType.entityGlint());
+            put(map2, RenderType.waterMask());
         });
         this.bufferSource = MultiBufferSource.immediateWithBuffers(map, new ByteBufferBuilder(786432));
         this.outlineBufferSource = new OutlineBufferSource(this.bufferSource);
