--- /dev/null
+++ b/net/caffeinemc/mods/sodium/client/gl/tessellation/GlTessellation.java
@@ -1,0 +_,13 @@
+package net.caffeinemc.mods.sodium.client.gl.tessellation;
+
+import net.caffeinemc.mods.sodium.client.gl.device.CommandList;
+
+public interface GlTessellation {
+    void delete(CommandList commandList);
+
+    void bind(CommandList commandList);
+
+    void unbind(CommandList commandList);
+
+    GlPrimitiveType getPrimitiveType();
+}
