--- /dev/null
+++ b/net/caffeinemc/mods/sodium/client/services/PlatformMixinOverrides.java
@@ -1,0 +_,17 @@
+package net.caffeinemc.mods.sodium.client.services;
+
+import java.util.List;
+
+public interface PlatformMixinOverrides {
+    PlatformMixinOverrides INSTANCE = Services.load(PlatformMixinOverrides.class);
+
+    static PlatformMixinOverrides getInstance() {
+        return INSTANCE;
+    }
+
+    List<MixinOverride> applyModOverrides();
+
+    record MixinOverride(String modId, String option, boolean enabled) {
+
+    }
+}
