--- a/net/minecraft/client/renderer/block/ModelBlockRenderer.java
+++ b/net/minecraft/client/renderer/block/ModelBlockRenderer.java
@@ -7,6 +_,7 @@
 import java.util.BitSet;
 import java.util.List;
 import javax.annotation.Nullable;
+
 import net.minecraft.CrashReport;
 import net.minecraft.CrashReportCategory;
 import net.minecraft.ReportedException;
@@ -31,23 +_,12 @@
     private final BlockColors blockColors;
     private static final int CACHE_SIZE = 100;
     static final ThreadLocal<ModelBlockRenderer.Cache> CACHE = ThreadLocal.withInitial(ModelBlockRenderer.Cache::new);
-
+    
     public ModelBlockRenderer(BlockColors blockColors) {
         this.blockColors = blockColors;
     }
 
-    public void tesselateBlock(
-        BlockAndTintGetter level,
-        BakedModel model,
-        BlockState state,
-        BlockPos pos,
-        PoseStack poseStack,
-        VertexConsumer consumer,
-        boolean checkSides,
-        RandomSource random,
-        long seed,
-        int packedOverlay
-    ) {
+    public void tesselateBlock(BlockAndTintGetter level, BakedModel model, BlockState state, BlockPos pos, PoseStack poseStack, VertexConsumer consumer, boolean checkSides, RandomSource random, long seed, int packedOverlay) {
         boolean flag = Minecraft.useAmbientOcclusion() && state.getLightEmission() == 0 && model.useAmbientOcclusion();
         poseStack.translate(state.getOffset(pos));
 
@@ -215,7 +_,7 @@
         );
     }
 
-    private void calculateShape(
+    public void calculateShape(
         BlockAndTintGetter level, BlockState state, BlockPos pos, int[] vertices, Direction direction, @Nullable float[] shape, BitSet shapeFlags
     ) {
         float f = 32.0F;
@@ -683,9 +_,9 @@
         }
     }
 
-    static class AmbientOcclusionFace {
-        final float[] brightness = new float[4];
-        final int[] lightmap = new int[4];
+    public static class AmbientOcclusionFace {
+        public final float[] brightness = new float[4];
+        public final int[] lightmap = new int[4];
 
         public AmbientOcclusionFace() {
         }
