--- /dev/null
+++ b/org/spongepowered/asm/mixin/injection/callback/Cancellable.java
@@ -1,0 +_,33 @@
+package org.spongepowered.asm.mixin.injection.callback;
+
+import java.util.concurrent.CancellationException;
+
+/**
+ * Interface for things which can be cancelled
+ */
+public interface Cancellable {
+
+    /**
+     * Get whether this is actually cancellable
+     *
+     * @return whether this is actually cancellable
+     */
+    public abstract boolean isCancellable();
+
+    /**
+     * Get whether this is cancelled
+     *
+     * @return whether this is cancelled
+     */
+    public abstract boolean isCancelled();
+
+    /**
+     * If the object is cancellable, cancels the object, implementors may throw
+     * an EventCancellationException if the object is not actually cancellable. 
+     *
+     * @throws CancellationException (optional) may be thrown if the object is
+     *      not actually cancellable. Contractually, this object may not throw
+     *      the exception if isCancellable() returns true.
+     */
+    public abstract void cancel() throws CancellationException;
+}
