--- a/net/minecraft/client/renderer/entity/ItemRenderer.java
+++ b/net/minecraft/client/renderer/entity/ItemRenderer.java
@@ -5,6 +_,7 @@
 import com.mojang.blaze3d.vertex.VertexConsumer;
 import com.mojang.blaze3d.vertex.VertexMultiConsumer;
 import com.mojang.math.MatrixUtil;
+
 import java.util.List;
 import javax.annotation.Nullable;
 import net.minecraft.client.Minecraft;
@@ -36,7 +_,7 @@
     public static final int NO_TINT = -1;
     private final ItemModelResolver resolver;
     private final ItemStackRenderState scratchItemStackRenderState = new ItemStackRenderState();
-
+    
     public ItemRenderer(ItemModelResolver resolver) {
         this.resolver = resolver;
     }
@@ -54,17 +_,7 @@
         renderQuadList(poseStack, buffer, model.getQuads(null, null, randomSource), tintLayers, packedLight, packedOverlay);
     }
 
-    public static void renderItem(
-        ItemDisplayContext displayContext,
-        PoseStack poseStack,
-        MultiBufferSource bufferSource,
-        int packedLight,
-        int packedOverlay,
-        int[] tintLayers,
-        BakedModel model,
-        RenderType renderType,
-        ItemStackRenderState.FoilType foilType
-    ) {
+    public static void renderItem(ItemDisplayContext displayContext, PoseStack poseStack, MultiBufferSource bufferSource, int packedLight, int packedOverlay, int[] tintLayers, BakedModel model, RenderType renderType, ItemStackRenderState.FoilType foilType) {
         VertexConsumer compassFoilBuffer;
         if (foilType == ItemStackRenderState.FoilType.SPECIAL) {
             PoseStack.Pose pose = poseStack.last().copy();
@@ -88,7 +_,7 @@
             : bufferSource.getBuffer(renderType);
     }
 
-    private static VertexConsumer getCompassFoilBuffer(MultiBufferSource bufferSource, RenderType renderType, PoseStack.Pose pose) {
+    public static VertexConsumer getCompassFoilBuffer(MultiBufferSource bufferSource, RenderType renderType, PoseStack.Pose pose) {
         return VertexMultiConsumer.create(
             new SheetedDecalTextureGenerator(bufferSource.getBuffer(RenderType.glint()), pose, 0.0078125F), bufferSource.getBuffer(renderType)
         );
