--- a/net/minecraft/world/level/FoliageColor.java
+++ b/net/minecraft/world/level/FoliageColor.java
@@ -5,7 +_,7 @@
     public static final int FOLIAGE_BIRCH = -8345771;
     public static final int FOLIAGE_DEFAULT = -12012264;
     public static final int FOLIAGE_MANGROVE = -7158200;
-    private static int[] pixels = new int[65536];
+    public static int[] pixels = new int[65536];
 
     public static void init(int[] foliageBuffer) {
         pixels = foliageBuffer;
