--- /dev/null
+++ b/net/caffeinemc/mods/sodium/client/gl/functions/DeviceFunctions.java
@@ -1,0 +_,15 @@
+package net.caffeinemc.mods.sodium.client.gl.functions;
+
+import net.caffeinemc.mods.sodium.client.gl.device.RenderDevice;
+
+public class DeviceFunctions {
+    private final BufferStorageFunctions bufferStorageFunctions;
+
+    public DeviceFunctions(RenderDevice device) {
+        this.bufferStorageFunctions = BufferStorageFunctions.pickBest(device);
+    }
+
+    public BufferStorageFunctions getBufferStorageFunctions() {
+        return this.bufferStorageFunctions;
+    }
+}
